### PR TITLE
build(895): Configure CircleCI to deploy to production

### LINKED ISSUE #895: Configure HHS/TANF-app:Master CircleCI jobs to deploy to the production space
We need to be able to trigger automated deployments to the production environment upon specific branch merges in https://github.com/HHS/TANF-app. When a merge is made from `HHS/TANF-app:main` to  `HHS/TANF-app:master`, a production deployment should be triggered, which includes the initial deployment of services infrastructure including S3 and RDS via Terraform, as well as the TANF application deployment to Cloud.gov. Additionally, a number of environment variables need to be added or updated in CircleCI and Cloud.gov.

AC:
- [x] Deploy latest code to production space on merges to `hhs:master` with CircleCI
- [x] Deploy services to production space with Terraform

Proposed tasks:
- [x] Expand [Terraform](https://github.com/raft-tech/TANF-app/tree/raft-tdp-main/terraform) manifest to include production deployment equivalent to `staging`
- [x] create `master` branch in `hhs:tanf-app`
- [x] confirm list of environment variables that should be added to circleci and cloud.gov
- [x]  Configure HHS:master CircleCI jobs to deploy to the production space
- [ ] Update ADR 009 (git-workflow) and ADR 008 (deployment flow) to reflect prod deployment to hhs:master (per #1731)

### PR DESCRIPTION
**Pull request by jorgegonzalez**
## Summary of Changes

This PR adds the production terraform configs and tweaks scripts to support the deployment to the tanf-prod space

Pull request closes #895

### TODO Upon Merging this PR

- [x] Update `ALLOWED_HOSTS` with prod backend route
- [x] Add following .env vars to the HHS CircleCI environment:
	- `CF_USERNAME_PROD`
	- `CF_PASSWORD_PROD`
	- `BASE_URL`
	- `FRONTEND_BASE_URL`
	- `DJANGO_SUPERUSER_NAME `
	- `DJANGO_SU_NAME`

**Note**: The existing service keys paired with `CF_USERNAME_PROD` and `CF_PASSWORD_PROD` were created for testing this PR in the Raft CircleCI env; these service keys need to be revoked and reissued for the HHS CircleCI.

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open http://localhost:3000/ and sign in.
1. Proceed with functional tests as described herein.
1. Test steps should be captured in the demo GIF(s) and/or screenshots below.
> *Demo GIF(s) and screenshots for testing procedure*

## Tabletop

- [x] Ready for Tabletop (Reviewer check this)
- [x] Tabletop scheduled (Assignee check this)
- [x] Tabletop approval 

## Deliverables
_More details on how deliverables herein are assessed included [here](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Deploy latest code to production space on merges to hhs:master with CircleCI
+ [x] Deploy services to production space with Terraform
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: terraform/production/variables.tf [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+variable "cf_api_url" {
+  type        = string
+  description = "cloud.gov api url"
+  default     = "https://api.fr.cloud.gov"
+}
+
+variable "cf_org_name" {
+  type        = string
+  description = "cloud.gov organization name"
+  default     = "hhs-acf-prototyping"
```
Comment (abottoms-coder) [inline terraform/production/variables.tf:16]: Because this is changing in a few days, should we go off id#? Or does this sufficiently get updated in a dynamic way and this default is less-used?

File: tdrs-backend/tdpservice/settings/cloudgov.py [hunk 2]
```diff
@@ -58,10 +58,14 @@ class CloudGov(Common):
     ###
     # Dynamic Database configuration based on cloud.gov services
     #
+    env_based_db_name = f'tdp_db_{cloudgov_space_suffix}_{cloudgov_name}'
+
+    db_name = database_creds['db_name'] if cloudgov_space_suffix == "prod" else  env_based_db_name
```
Comment (abottoms-coder) [inline tdrs-backend/tdpservice/settings/cloudgov.py]: This would use the auto-generated db name 'cw....', right?  Were you unable or didn't want to manually create that database?

Comment (jorgegonzalez) [reply tdrs-backend/tdpservice/settings/cloudgov.py]: Not sure why it had to be manually created if no other env is sharing this instance?

Comment (abottoms-coder) [reply tdrs-backend/tdpservice/settings/cloudgov.py]: It doesn't have to be created, just noticed it was different than the other environments for naming convention. Not important to fix, fine as-is

File: scripts/deploy-backend.sh [hunk 3]
```diff
@@ -27,9 +30,10 @@ generate_jwt_cert()
 
 update_backend()
 {
-    bash ./scripts/set-backend-env-vars.sh "$CGHOSTNAME_BACKEND" "$CF_SPACE"
     cd tdrs-backend || exit
     if [ "$1" = "rolling" ] ; then
+          bash ./scripts/set-backend-env-vars.sh "$CGHOSTNAME_BACKEND" "$CF_SPACE"
```
Comment (abottoms-coder) [inline scripts/deploy-backend.sh:35]: Can you elaborate on why this was moved back? I was seeing bad behavior with `rebuild` strategy deployments. We can address during tabletop if preferable.

File: .circleci/config.yml [hunk 4]
```diff
@@ -583,15 +601,6 @@ workflows:
           requires:
             - secrets-check
 
-  dev-deployment:
```
Comment (abottoms-coder) [inline .circleci/config.yml:586]: Was this removal intentional?

File: scripts/deploy-backend.sh (hunk 3 above)
Comment (jorgegonzalez) [reply scripts/deploy-backend.sh:35]: Addressed in tabletop, note that this can be reverted later potentially

**Review by abottoms-coder — APPROVED** [review]

**Review by raftmsohani — APPROVED** [review]
LGTM! For database backup, we need to add APT package to buildpacks to be able to install custom packages. There are different ways to do it, one which was to add it to the CircleCI script, but even without touching this script, we should still be able to install custom packages. So everything is ok

**Review by ADPennington — APPROVED** [review]
lgtm @jorgegonzalez 🚀 I updated hhs ci project settings with appropriate EVs. Note that all of the needed env vars have not yet been set yet for prod (e.g. nextgen), so I've re-opened #897 to track that. 

@abottoms-coder can proceed with updating `allowed_hosts`.
